Find SIRF_PATH based on py install dir [ci skip]

ashgillman opened the pull request:
In my installation I don't usually bother with $SIRF_PATH. I think we could try and guess it from the Python directory?

This works for me locally.

evgueni-ovtchinnikov commented on src/common/Utilities.py:
```diff
@@ -55,9 +55,32 @@ def examples_data_path(data_type):
+    try:
+        py_root = sirf.__file__
+        try:
+            # is install dir like $SIRF_PATH/lib/python3.8/site-packages?
```
should be SIRF_INSTALL_PATH

evgueni-ovtchinnikov commented on src/common/Utilities.py:
```diff
@@ -55,9 +55,32 @@ def examples_data_path(data_type):
+            # is install dir like $SIRF_PATH/lib/python3.8/site-packages?
+            lib_idx = py_root.split(os.sep).index('lib')
+        except ValueError:
+            # is install dir like $SIRF_PATH/python?
```
as above

evgueni-ovtchinnikov commented on src/common/Utilities.py:
```diff
@@ -55,9 +55,32 @@ def examples_data_path(data_type):
+        except ValueError:
+            # is install dir like $SIRF_PATH/python?
+            lib_idx = py_root.split(os.sep).index('python')
+        potential_install_loc = os.path.join(os.sep, *py_root.split(os.sep)[:lib_idx])
```
wrong path on Windows (\ missing after C:)

ashgillman replied:
Ah, thanks for checking

ashgillman commented:
@evgueni-ovtchinnikov - According to StackOverflow the latest change should fix the abs path issue on Windows but I'm unable to confirm. Can you please check?

ashgillman commented on src/common/Utilities.py:
```diff
@@ -60,12 +60,13 @@ def examples_data_path(data_type):
+            # is install dir like $SIRF_INSTALL_PATH/python?
             lib_idx = py_root.split(os.sep).index('python')
-        potential_install_loc = os.path.join(os.sep, *py_root.split(os.sep)[:lib_idx])
+        potential_install_loc = os.path.abspath(
+            os.path.join(os.sep, *py_root.split(os.sep)[:lib_idx]))
```
Hmm, I'm confused. I don't have SIRF installed on Windows to replicate, but I've tried a minimum example.

On Windows:
```python
>>> import os                                                                                                                     
>>> py_root = os.__file__                                                                                                         
>>> lib_ids = py_root.split(os.sep).index('lib')                                                                                  
>>> lib_idx = py_root.split(os.sep).index('lib')                                                                                  
>>> potential_install_loc = os.path.abspath(os.path.join(os.sep, *py_root.split(os.sep)[:lib_idx]))                               
>>> print(potential_install_loc)                                                                                                  
C:\Windows\system32\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.1520.0_x64__qbz5n2kfra8p0 
```

On Linux:
```python
In [7]: import os                                                                               
In [8]: py_root = os.__file__                                                                   
In [9]: lib_idx = py_root.split(os.sep).index('lib')                                            
In [10]: potential_install_loc = os.path.abspath(os.path.join(os.sep, *py_root.split(os.sep)[:li
    ...: b_idx]))                                                                               
In [11]: print(potential_install_loc)                                                           
/home/gil2a4/anaconda3
```

So I don't get the repetition that you get. Indeed, a quick Google suggests that `os.path.abspath(os.path.join(os.sep, *PATH))` is a [recommended way of getting the absolute version of PATH on Windows and Linux](https://stackoverflow.com/a/28080468)

evgueni-ovtchinnikov replied:
Tried your Windows example, got this:
```
C:\Users\wps46139\Documents\GitHub\build\sources\SIRF\examples\Python\PET\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64
```

evgueni-ovtchinnikov replied:
how about this (on Windows):
```
root_split = sirf.__file__.split(os.sep)
idx = root_split.index('python')
potential_install_loc = root_split[0] + os.path.join(os.sep, root_split[1 : idx])
```

ashgillman replied:
@evgueni-ovtchinnikov  - that seemed to work for me. On Windows root_split[0] is the drive, and on Unix it is empty, so this satisfies both. This seems kind-of "cowboy", but I've given this some diligence and I'm surprised to no find much!

Probably there is something more elegant using pathlib, but since that's not used elsewhere this would be more rather than less confusing for other developers?

Anyway, check out the latest commit below

KrisThielemans commented:
I haven't checked the code for any of this but note that `os.path.abspath(os.path.join(os.sep, *py_roo...))` is suspicious as on Windows (or VMS!) that starts with `os.sep` can be very surprising indeed (if your current working dir is on a differnt drive, or a network drive etc).

Also, I'd hate to rely on `python` by in the name (`ipython`, `cython`, `pypy` etc...)

ashgillman commented:
> I haven't checked the code for any of this but note that `os.path.abspath(os.path.join(os.sep, *py_roo...))` is suspicious as on Windows (or VMS!) that starts with `os.sep` can be very surprising indeed (if your current working dir is on a differnt drive, or a network drive etc).

I think that first sep is just ignored on Windows.
Note that the first element of `py_root` is 'C:', but the join unintuitively and inexplicably omits the separator after 'C:':

```
>>> os.__file__
'C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.9_3.9.1520.0_x64__qbz5n2kfra8p0\\lib\\os.py'
>>> os.__file__.split(os.sep)
['C:', 'Program Files', 'WindowsApps', 'PythonSoftwareFoundation.Python.3.9_3.9.1520.0_x64__qbz5n2kfra8p0', 'lib', 'os.py']
>>> os.path.join(*os.__file__.split(os.sep))
'C:Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.9_3.9.1520.0_x64__qbz5n2kfra8p0\\lib\\os.py'
```

> Also, I'd hate to rely on `python` by in the name (`ipython`, `cython`, `pypy` etc...)

This was just finding the default install directory of `<install_loc>/python` used with CMake. So this default should account for people using the SuperBuild or installing to a default Python location, so should suit 90% of people so I hope a reasonable default search location.

evgueni-ovtchinnikov commented on src/common/Utilities.py:
```diff
@@ -55,9 +55,41 @@ def examples_data_path(data_type):
+    # one final effort
+    try:
+        # Try find <SIRF_INSTALL_PATH>/share/data/examples
+        py_root = os.path.abspath(sirf.__file__)
```
if I am not mistaken, you never use `py_root` except in `py_root.split(os.sep)` - I suggest replacing this with
```
py_root_split = os.path.abspath(sirf.__file__).split(os.sep)
```

KrisThielemans commented:
I think we can resolve this via #1202 and #1201

evgueni-ovtchinnikov commented:
made obsolete by #1201, closing
